swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -399,6 +399,16 @@ private module ReturnNodes {
     override ReturnKind getKind() { result instanceof NormalReturnKind }
   }
 
+  /**
+   * A data-flow node that represents the `self` value in a constructor being
+   * implicitly returned as the newly-constructed object
+   * */
+  class SelfReturnNode extends InoutReturnNodeImpl {
+    SelfReturnNode() { exit.getScope() instanceof ConstructorDecl }
+
+    override ReturnKind getKind() { result instanceof NormalReturnKind }
+  }
```

Comment:
The refactor you did in https://github.com/github/codeql/pull/12415/commits/86fd21de452bb83850fea9182b3bbd2d78876e7a threw away the constraint that the parameter must be a `SelfParamDecl`. I think we still need this to not have all parameters exit via a `NormalReturnKind`.

Reply:
Good spot.  I'm a little surprised there are no spurious test results though!

Reply:
Yes. It would require a test with a constructor that writes to an inout parameter, which we might not have currently. Something like:
```swift
class X {
  init(_ n : inout Int) { n = source() }
}
func foo() {
  var n = 0
  sink(X(&n))
}
```
i.e., we write to a non-self parameter, but because we're forgetting the `SelfParamDecl` constraint this non-self parameter is leaving the function as the return value.